// sift/src/core.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***
exports.createQueryTester = exports.createOperationTester = exports.createQueryOperation = exports.containsOperation = exports.numericalOperation = exports.numericalOperationCreator = exports.NopeOperation = exports.createEqualsOperation = exports.EqualsOperation = exports.createTester = exports.NestedOperation = exports.QueryOperation = exports.NamedGroupOperation = exports.BaseOperation = void 0;
const utils_1 = require("./utils");
/**
 * Walks through each value given the context - used for nested operations. E.g:
 * { "person.address": { $eq: "blarg" }}
 */
const walkKeyPathValues = (item, keyPath, next, depth, key, owner) => {
    const currentKey = keyPath[depth];
    // if array, then try matching. Might fall through for cases like:
    // { $eq: [1, 2, 3] }, [ 1, 2, 3 ].
    if ((0, utils_1.isArray)(item) && isNaN(Number(currentKey))) {
        for (let i = 0, { length } = item; i < length; i++) {
            // if FALSE is returned, then terminate walker. For operations, this simply
            // means that the search critera was met.
            if (!walkKeyPathValues(item[i], keyPath, next, depth, i, item)) {
                return false;
        ***REMOVED***
    ***REMOVED***
***REMOVED***
    if (depth === keyPath.length || item == null) {
        return next(item, key, owner, depth === 0);
***REMOVED***
    return walkKeyPathValues(item[currentKey], keyPath, next, depth + 1, currentKey, item);
};
class BaseOperation {
    constructor(params, owneryQuery, options, name) {
        this.params = params;
        this.owneryQuery = owneryQuery;
        this.options = options;
        this.name = name;
        this.init();
***REMOVED***
    init() { }
    reset() {
        this.done = false;
        this.keep = false;
***REMOVED***
}
exports.BaseOperation = BaseOperation;
class GroupOperation extends BaseOperation {
    constructor(params, owneryQuery, options, children) {
        super(params, owneryQuery, options);
        this.children = children;
***REMOVED***
    /**
     */
    reset() {
        this.keep = false;
        this.done = false;
        for (let i = 0, { length } = this.children; i < length; i++) {
            this.children[i].reset();
    ***REMOVED***
***REMOVED***
    /**
     */
    childrenNext(item, key, owner, root) {
        let done = true;
        let keep = true;
        for (let i = 0, { length } = this.children; i < length; i++) {
            const childOperation = this.children[i];
            if (!childOperation.done) {
                childOperation.next(item, key, owner, root);
        ***REMOVED***
            if (!childOperation.keep) {
                keep = false;
        ***REMOVED***
            if (childOperation.done) {
                if (!childOperation.keep) {
                    break;
            ***REMOVED***
        ***REMOVED***
            else {
                done = false;
        ***REMOVED***
    ***REMOVED***
        this.done = done;
        this.keep = keep;
***REMOVED***
}
class NamedGroupOperation extends GroupOperation {
    constructor(params, owneryQuery, options, children, name) {
        super(params, owneryQuery, options, children);
        this.name = name;
***REMOVED***
}
exports.NamedGroupOperation = NamedGroupOperation;
class QueryOperation extends GroupOperation {
    constructor() {
        super(...arguments);
        this.propop = true;
***REMOVED***
    /**
     */
    next(item, key, parent, root) {
        this.childrenNext(item, key, parent, root);
***REMOVED***
}
exports.QueryOperation = QueryOperation;
class NestedOperation extends GroupOperation {
    constructor(keyPath, params, owneryQuery, options, children) {
        super(params, owneryQuery, options, children);
        this.keyPath = keyPath;
        this.propop = true;
        /**
         */
        this._nextNestedValue = (value, key, owner, root) => {
            this.childrenNext(value, key, owner, root);
            return !this.done;
    ***REMOVED***;
***REMOVED***
    /**
     */
    next(item, key, parent) {
        walkKeyPathValues(item, this.keyPath, this._nextNestedValue, 0, key, parent);
***REMOVED***
}
exports.NestedOperation = NestedOperation;
const createTester = (a, compare) => {
    if (a instanceof Function) {
        return a;
***REMOVED***
    if (a instanceof RegExp) {
        return b => {
            const result = typeof b === "string" && a.test(b);
            a.lastIndex = 0;
            return result;
    ***REMOVED***;
***REMOVED***
    const comparableA = (0, utils_1.comparable)(a);
    return b => compare(comparableA, (0, utils_1.comparable)(b));
};
exports.createTester = createTester;
class EqualsOperation extends BaseOperation {
    constructor() {
        super(...arguments);
        this.propop = true;
***REMOVED***
    init() {
        this._test = (0, exports.createTester)(this.params, this.options.compare);
***REMOVED***
    next(item, key, parent) {
        if (!Array.isArray(parent) || parent.hasOwnProperty(key)) {
            if (this._test(item, key, parent)) {
                this.done = true;
                this.keep = true;
        ***REMOVED***
    ***REMOVED***
***REMOVED***
}
exports.EqualsOperation = EqualsOperation;
const createEqualsOperation = (params, owneryQuery, options) => new EqualsOperation(params, owneryQuery, options);
exports.createEqualsOperation = createEqualsOperation;
class NopeOperation extends BaseOperation {
    constructor() {
        super(...arguments);
        this.propop = true;
***REMOVED***
    next() {
        this.done = true;
        this.keep = false;
***REMOVED***
}
exports.NopeOperation = NopeOperation;
const numericalOperationCreator = (createNumericalOperation) => (params, owneryQuery, options, name) => {
    if (params == null) {
        return new NopeOperation(params, owneryQuery, options, name);
***REMOVED***
    return createNumericalOperation(params, owneryQuery, options, name);
};
exports.numericalOperationCreator = numericalOperationCreator;
const numericalOperation = (createTester) => (0, exports.numericalOperationCreator)((params, owneryQuery, options, name) => {
    const typeofParams = typeof (0, utils_1.comparable)(params);
    const test = createTester(params);
    return new EqualsOperation(b => {
        return typeof (0, utils_1.comparable)(b) === typeofParams && test(b);
***REMOVED***, owneryQuery, options, name);
***REMOVED***
exports.numericalOperation = numericalOperation;
const createNamedOperation = (name, params, parentQuery, options) => {
    const operationCreator = options.operations[name];
    if (!operationCreator) {
        throwUnsupportedOperation(name);
***REMOVED***
    return operationCreator(params, parentQuery, options, name);
};
const throwUnsupportedOperation = (name) => {
    throw new Error(`Unsupported operation: ${name}`);
};
const containsOperation = (query, options) => {
    for (const key in query) {
        if (options.operations.hasOwnProperty(key) || key.charAt(0) === "$")
            return true;
***REMOVED***
    return false;
};
exports.containsOperation = containsOperation;
const createNestedOperation = (keyPath, nestedQuery, parentKey, owneryQuery, options) => {
    if ((0, exports.containsOperation)(nestedQuery, options)) {
        const [selfOperations, nestedOperations] = createQueryOperations(nestedQuery, parentKey, options);
        if (nestedOperations.length) {
            throw new Error(`Property queries must contain only operations, or exact objects.`);
    ***REMOVED***
        return new NestedOperation(keyPath, nestedQuery, owneryQuery, options, selfOperations);
***REMOVED***
    return new NestedOperation(keyPath, nestedQuery, owneryQuery, options, [
        new EqualsOperation(nestedQuery, owneryQuery, options)
    ]);
};
const createQueryOperation = (query, owneryQuery = null, { compare, operations } = {}) => {
    const options = {
        compare: compare || utils_1.equals,
        operations: Object.assign({}, operations || {})
***REMOVED***;
    const [selfOperations, nestedOperations] = createQueryOperations(query, null, options);
    const ops = [];
    if (selfOperations.length) {
        ops.push(new NestedOperation([], query, owneryQuery, options, selfOperations));
***REMOVED***
    ops.push(...nestedOperations);
    if (ops.length === 1) {
        return ops[0];
***REMOVED***
    return new QueryOperation(query, owneryQuery, options, ops);
};
exports.createQueryOperation = createQueryOperation;
const createQueryOperations = (query, parentKey, options) => {
    const selfOperations = [];
    const nestedOperations = [];
    if (!(0, utils_1.isVanillaObject)(query)) {
        selfOperations.push(new EqualsOperation(query, query, options));
        return [selfOperations, nestedOperations];
***REMOVED***
    for (const key in query) {
        if (options.operations.hasOwnProperty(key)) {
            const op = createNamedOperation(key, query[key], query, options);
            if (op) {
                if (!op.propop && parentKey && !options.operations[parentKey]) {
                    throw new Error(`Malformed query. ${key} cannot be matched against property.`);
            ***REMOVED***
        ***REMOVED***
            // probably just a flag for another operation (like $options)
            if (op != null) {
                selfOperations.push(op);
        ***REMOVED***
    ***REMOVED***
        else if (key.charAt(0) === "$") {
            throwUnsupportedOperation(key);
    ***REMOVED***
        else {
            nestedOperations.push(createNestedOperation(key.split("."), query[key], key, query, options));
    ***REMOVED***
***REMOVED***
    return [selfOperations, nestedOperations];
};
const createOperationTester = (operation) => (item, key, owner) => {
    operation.reset();
    operation.next(item, key, owner);
    return operation.keep;
};
exports.createOperationTester = createOperationTester;
const createQueryTester = (query, options = {}) => {
    return (0, exports.createOperationTester)((0, exports.createQueryOperation)(query, null, options));
};
exports.createQueryTester = createQueryTester;
