// sift/src/utils.ts
export type Key = string | number;
export type Comparator = (a, b) => boolean;
export const typeChecker = <TType>(type) => {
  const typeString = "[object " + type + "]";
  return function(value): value is TType {
    return getClassName(value) === typeString;
***REMOVED***;
};

const getClassName = value => Object.prototype.toString.call(value);

export const comparable = (value: any) => {
  if (value instanceof Date) {
    return value.getTime();
***REMOVED*** else if (isArray(value)) {
    return value.map(comparable);
***REMOVED*** else if (value && typeof value.toJSON === "function") {
    return value.toJSON();
***REMOVED***

  return value;
};

export const isArray = typeChecker<Array<any>>("Array");
export const isObject = typeChecker<Object>("Object");
export const isFunction = typeChecker<Function>("Function");
export const isVanillaObject = value => {
  return (
    value &&
    (value.constructor === Object ||
      value.constructor === Array ||
      value.constructor.toString() === "function Object() { [native code] }" ||
      value.constructor.toString() === "function Array() { [native code] }") &&
    !value.toJSON
  );
};

export const equals = (a, b) => {
  if (a == null && a == b) {
    return true;
***REMOVED***
  if (a === b) {
    return true;
***REMOVED***

  if (Object.prototype.toString.call(a) !== Object.prototype.toString.call(b)) {
    return false;
***REMOVED***

  if (isArray(a)) {
    if (a.length !== b.length) {
      return false;
***REMOVED***
    for (let i = 0, { length } = a; i < length; i++) {
      if (!equals(a[i], b[i])) return false;
***REMOVED***
    return true;
***REMOVED*** else if (isObject(a)) {
    if (Object.keys(a).length !== Object.keys(b).length) {
      return false;
***REMOVED***
    for (const key in a) {
      if (!equals(a[key], b[key])) return false;
***REMOVED***
    return true;
***REMOVED***
  return false;
};
