// mongoose/lib/plugins/sharding.js
'use strict';

const objectIdSymbol = require('../helpers/symbols').objectIdSymbol;
const utils = require('../utils');

/*!
 * ignore
 */

module.exports = function shardingPlugin(schema) {
  schema.post('init', function shardingPluginPostInit() {
    storeShard.call(this);
    return this;
  ***REMOVED***
  schema.pre('save', function shardingPluginPreSave(next) {
    applyWhere.call(this);
    next();
  ***REMOVED***
  schema.pre('remove', function shardingPluginPreRemove(next) {
    applyWhere.call(this);
    next();
  ***REMOVED***
  schema.post('save', function shardingPluginPostSave() {
    storeShard.call(this);
  ***REMOVED***
};

/*!
 * ignore
 */

function applyWhere() {
  let paths;
  let len;

  if (this.$__.shardval) {
    paths = Object.keys(this.$__.shardval);
    len = paths.length;

    this.$where = this.$where || {};
    for (let i = 0; i < len; ++i) {
      this.$where[paths[i]] = this.$__.shardval[paths[i]];
***REMOVED***
***REMOVED***
}

/*!
 * ignore
 */

module.exports.storeShard = storeShard;

/*!
 * ignore
 */

function storeShard() {
  // backwards compat
  const key = this.$__schema.options.shardKey || this.$__schema.options.shardkey;
  if (!utils.isPOJO(key)) {
    return;
***REMOVED***

  const orig = this.$__.shardval = {};
  const paths = Object.keys(key);
  const len = paths.length;
  let val;

  for (let i = 0; i < len; ++i) {
    val = this.$__getValue(paths[i]);
    if (val == null) {
      orig[paths[i]] = val;
***REMOVED*** else if (utils.isMongooseObject(val)) {
      orig[paths[i]] = val.toObject({ depopulate: true, _isNested: true ***REMOVED***
***REMOVED*** else if (val instanceof Date || val[objectIdSymbol]) {
      orig[paths[i]] = val;
***REMOVED*** else if (typeof val.valueOf === 'function') {
      orig[paths[i]] = val.valueOf();
***REMOVED***
      orig[paths[i]] = val;
***REMOVED***
***REMOVED***
}
