// @aws-sdk/client-sso/dist-es/commands/LogoutCommand.js
import { getEndpointPlugin } from "@aws-sdk/middleware-endpoint";
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { LogoutRequestFilterSensitiveLog } from "../models/models_0";
import { deserializeAws_restJson1LogoutCommand, serializeAws_restJson1LogoutCommand } from "../protocols/Aws_restJson1";
export class LogoutCommand extends $Command {
    constructor(input) {
        super();
        this.input = input;
***REMOVED***
    static getEndpointParameterInstructions() {
        return {
            UseFIPS: { type: "builtInParams", name: "useFipsEndpoint" },
            Endpoint: { type: "builtInParams", name: "endpoint" },
            Region: { type: "builtInParams", name: "region" },
            UseDualStack: { type: "builtInParams", name: "useDualstackEndpoint" },
    ***REMOVED***;
***REMOVED***
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        this.middlewareStack.use(getEndpointPlugin(configuration, LogoutCommand.getEndpointParameterInstructions()));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "SSOClient";
        const commandName = "LogoutCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: LogoutRequestFilterSensitiveLog,
            outputFilterSensitiveLog: (output) => output,
    ***REMOVED***;
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
***REMOVED***
    serialize(input, context) {
        return serializeAws_restJson1LogoutCommand(input, context);
***REMOVED***
    deserialize(output, context) {
        return deserializeAws_restJson1LogoutCommand(output, context);
***REMOVED***
}
