// mongodb/lib/operations/add_user.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***
exports.AddUserOperation = void 0;
const crypto = require("crypto");
const error_1 = require("../error");
const utils_1 = require("../utils");
const command_1 = require("./command");
const operation_1 = require("./operation");
/** @internal */
class AddUserOperation extends command_1.CommandOperation {
    constructor(db, username, password, options) {
        super(db, options);
        this.db = db;
        this.username = username;
        this.password = password;
        this.options = options !== null && options !== void 0 ? options : {};
***REMOVED***
    execute(server, session, callback) {
        const db = this.db;
        const username = this.username;
        const password = this.password;
        const options = this.options;
        // Error out if digestPassword set
        if (options.digestPassword != null) {
            return callback(new error_1.MongoInvalidArgumentError('Option "digestPassword" not supported via addUser, use db.command(...) instead'));
    ***REMOVED***
        let roles;
        if (!options.roles || (Array.isArray(options.roles) && options.roles.length === 0)) {
            (0, utils_1.emitWarningOnce)('Creating a user without roles is deprecated. Defaults to "root" if db is "admin" or "dbOwner" otherwise');
            if (db.databaseName.toLowerCase() === 'admin') {
                roles = ['root'];
        ***REMOVED***
            else {
                roles = ['dbOwner'];
        ***REMOVED***
    ***REMOVED***
        else {
            roles = Array.isArray(options.roles) ? options.roles : [options.roles];
    ***REMOVED***
        let topology;
      ***REMOVED***
            topology = (0, utils_1.getTopology)(db);
    ***REMOVED***
        catch (error) {
            return callback(error);
    ***REMOVED***
        const digestPassword = topology.lastHello().maxWireVersion >= 7;
        let userPassword = password;
        if (!digestPassword) {
            // Use node md5 generator
            const md5 = crypto.createHash('md5');
            // Generate keys used for authentication
            md5.update(`${username}:mongo:${password}`);
            userPassword = md5.digest('hex');
    ***REMOVED***
        // Build the command to execute
        const command = {
            createUser: username,
            customData: options.customData || {},
            roles: roles,
            digestPassword
    ***REMOVED***;
        // No password
        if (typeof password === 'string') {
            command.pwd = userPassword;
    ***REMOVED***
        super.executeCommand(server, session, command, callback);
***REMOVED***
}
exports.AddUserOperation = AddUserOperation;
(0, operation_1.defineAspects)(AddUserOperation, [operation_1.Aspect.WRITE_OPERATION]);
