// socks/build/common/receivebuffer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***
exports.ReceiveBuffer = void 0;
class ReceiveBuffer {
    constructor(size = 4096) {
        this.buffer = Buffer.allocUnsafe(size);
        this.offset = 0;
        this.originalSize = size;
***REMOVED***
    get length() {
        return this.offset;
***REMOVED***
    append(data) {
        if (!Buffer.isBuffer(data)) {
            throw new Error('Attempted to append a non-buffer instance to ReceiveBuffer.');
    ***REMOVED***
        if (this.offset + data.length >= this.buffer.length) {
            const tmp = this.buffer;
            this.buffer = Buffer.allocUnsafe(Math.max(this.buffer.length + this.originalSize, this.buffer.length + data.length));
            tmp.copy(this.buffer);
    ***REMOVED***
        data.copy(this.buffer, this.offset);
        return (this.offset += data.length);
***REMOVED***
    peek(length) {
        if (length > this.offset) {
            throw new Error('Attempted to read beyond the bounds of the managed internal data.');
    ***REMOVED***
        return this.buffer.slice(0, length);
***REMOVED***
    get(length) {
        if (length > this.offset) {
            throw new Error('Attempted to read beyond the bounds of the managed internal data.');
    ***REMOVED***
        const value = Buffer.allocUnsafe(length);
        this.buffer.slice(0, length).copy(value);
        this.buffer.copyWithin(0, length, length + this.offset - length);
        this.offset -= length;
        return value;
***REMOVED***
}
exports.ReceiveBuffer = ReceiveBuffer;
