// bson/src/symbol.ts
/** @public */
export interface BSONSymbolExtended {
  $symbol: string;
}

/**
 * A class representation of the BSON Symbol type.
 * @public
 * @category BSONType
 */
export class BSONSymbol {
  _bsontype!: 'Symbol';

  value!: string;
  /**
   * @param value - the string representing the symbol.
   */
  constructor(value: string) {
    if (!(this instanceof BSONSymbol)) return new BSONSymbol(value);

    this.value = value;
***REMOVED***

  /** Access the wrapped string value. */
  valueOf(): string {
    return this.value;
***REMOVED***

  toString(): string {
    return this.value;
***REMOVED***

  /** @internal */
  inspect(): string {
    return `new BSONSymbol("${this.value}")`;
***REMOVED***

  toJSON(): string {
    return this.value;
***REMOVED***

  /** @internal */
  toExtendedJSON(): BSONSymbolExtended {
    return { $symbol: this.value };
***REMOVED***

  /** @internal */
  static fromExtendedJSON(doc: BSONSymbolExtended): BSONSymbol {
    return new BSONSymbol(doc.$symbol);
***REMOVED***

  /** @internal */
  [Symbol.for('nodejs.util.inspect.custom')](): string {
    return this.inspect();
***REMOVED***
}

Object.defineProperty(BSONSymbol.prototype, '_bsontype', { value: 'Symbol' ***REMOVED***
