// sift/es/index.js
const typeChecker = (type) => {
    const typeString = "[object " + type + "]";
    return function (value) {
        return getClassName(value) === typeString;
***REMOVED***;
};
const getClassName = value => Object.prototype.toString.call(value);
const comparable = (value) => {
    if (value instanceof Date) {
        return value.getTime();
***REMOVED***
    else if (isArray(value)) {
        return value.map(comparable);
***REMOVED***
    else if (value && typeof value.toJSON === "function") {
        return value.toJSON();
***REMOVED***
    return value;
};
const isArray = typeChecker("Array");
const isObject = typeChecker("Object");
const isFunction = typeChecker("Function");
const isVanillaObject = value => {
    return (value &&
        (value.constructor === Object ||
            value.constructor === Array ||
            value.constructor.toString() === "function Object() { [native code] }" ||
            value.constructor.toString() === "function Array() { [native code] }") &&
        !value.toJSON);
};
const equals = (a, b) => {
    if (a == null && a == b) {
        return true;
***REMOVED***
    if (a === b) {
        return true;
***REMOVED***
    if (Object.prototype.toString.call(a) !== Object.prototype.toString.call(b)) {
        return false;
***REMOVED***
    if (isArray(a)) {
        if (a.length !== b.length) {
            return false;
    ***REMOVED***
        for (let i = 0, { length } = a; i < length; i++) {
            if (!equals(a[i], b[i]))
                return false;
    ***REMOVED***
        return true;
***REMOVED***
    else if (isObject(a)) {
        if (Object.keys(a).length !== Object.keys(b).length) {
            return false;
    ***REMOVED***
        for (const key in a) {
            if (!equals(a[key], b[key]))
                return false;
    ***REMOVED***
        return true;
***REMOVED***
    return false;
};

/**
 * Walks through each value given the context - used for nested operations. E.g:
 * { "person.address": { $eq: "blarg" }}
 */
const walkKeyPathValues = (item, keyPath, next, depth, key, owner) => {
    const currentKey = keyPath[depth];
    // if array, then try matching. Might fall through for cases like:
    // { $eq: [1, 2, 3] }, [ 1, 2, 3 ].
    if (isArray(item) && isNaN(Number(currentKey))) {
        for (let i = 0, { length } = item; i < length; i++) {
            // if FALSE is returned, then terminate walker. For operations, this simply
            // means that the search critera was met.
            if (!walkKeyPathValues(item[i], keyPath, next, depth, i, item)) {
                return false;
        ***REMOVED***
    ***REMOVED***
***REMOVED***
    if (depth === keyPath.length || item == null) {
        return next(item, key, owner, depth === 0);
***REMOVED***
    return walkKeyPathValues(item[currentKey], keyPath, next, depth + 1, currentKey, item);
};
class BaseOperation {
    constructor(params, owneryQuery, options, name) {
        this.params = params;
        this.owneryQuery = owneryQuery;
        this.options = options;
        this.name = name;
        this.init();
***REMOVED***
    init() { }
    reset() {
        this.done = false;
        this.keep = false;
***REMOVED***
}
class GroupOperation extends BaseOperation {
    constructor(params, owneryQuery, options, children) {
        super(params, owneryQuery, options);
        this.children = children;
***REMOVED***
    /**
     */
    reset() {
        this.keep = false;
        this.done = false;
        for (let i = 0, { length } = this.children; i < length; i++) {
            this.children[i].reset();
    ***REMOVED***
***REMOVED***
    /**
     */
    childrenNext(item, key, owner, root) {
        let done = true;
        let keep = true;
        for (let i = 0, { length } = this.children; i < length; i++) {
            const childOperation = this.children[i];
            if (!childOperation.done) {
                childOperation.next(item, key, owner, root);
        ***REMOVED***
            if (!childOperation.keep) {
                keep = false;
        ***REMOVED***
            if (childOperation.done) {
                if (!childOperation.keep) {
                    break;
            ***REMOVED***
        ***REMOVED***
            else {
                done = false;
        ***REMOVED***
    ***REMOVED***
        this.done = done;
        this.keep = keep;
***REMOVED***
}
class NamedGroupOperation extends GroupOperation {
    constructor(params, owneryQuery, options, children, name) {
        super(params, owneryQuery, options, children);
        this.name = name;
***REMOVED***
}
class QueryOperation extends GroupOperation {
    constructor() {
        super(...arguments);
        this.propop = true;
***REMOVED***
    /**
     */
    next(item, key, parent, root) {
        this.childrenNext(item, key, parent, root);
***REMOVED***
}
class NestedOperation extends GroupOperation {
    constructor(keyPath, params, owneryQuery, options, children) {
        super(params, owneryQuery, options, children);
        this.keyPath = keyPath;
        this.propop = true;
        /**
         */
        this._nextNestedValue = (value, key, owner, root) => {
            this.childrenNext(value, key, owner, root);
            return !this.done;
    ***REMOVED***;
***REMOVED***
    /**
     */
    next(item, key, parent) {
        walkKeyPathValues(item, this.keyPath, this._nextNestedValue, 0, key, parent);
***REMOVED***
}
const createTester = (a, compare) => {
    if (a instanceof Function) {
        return a;
***REMOVED***
    if (a instanceof RegExp) {
        return b => {
            const result = typeof b === "string" && a.test(b);
            a.lastIndex = 0;
            return result;
    ***REMOVED***;
***REMOVED***
    const comparableA = comparable(a);
    return b => compare(comparableA, comparable(b));
};
class EqualsOperation extends BaseOperation {
    constructor() {
        super(...arguments);
        this.propop = true;
***REMOVED***
    init() {
        this._test = createTester(this.params, this.options.compare);
***REMOVED***
    next(item, key, parent) {
        if (!Array.isArray(parent) || parent.hasOwnProperty(key)) {
            if (this._test(item, key, parent)) {
                this.done = true;
                this.keep = true;
        ***REMOVED***
    ***REMOVED***
***REMOVED***
}
const createEqualsOperation = (params, owneryQuery, options) => new EqualsOperation(params, owneryQuery, options);
class NopeOperation extends BaseOperation {
    constructor() {
        super(...arguments);
        this.propop = true;
***REMOVED***
    next() {
        this.done = true;
        this.keep = false;
***REMOVED***
}
const numericalOperationCreator = (createNumericalOperation) => (params, owneryQuery, options, name) => {
    if (params == null) {
        return new NopeOperation(params, owneryQuery, options, name);
***REMOVED***
    return createNumericalOperation(params, owneryQuery, options, name);
};
const numericalOperation = (createTester) => numericalOperationCreator((params, owneryQuery, options, name) => {
    const typeofParams = typeof comparable(params);
    const test = createTester(params);
    return new EqualsOperation(b => {
        return typeof comparable(b) === typeofParams && test(b);
***REMOVED***, owneryQuery, options, name);
***REMOVED***
const createNamedOperation = (name, params, parentQuery, options) => {
    const operationCreator = options.operations[name];
    if (!operationCreator) {
        throwUnsupportedOperation(name);
***REMOVED***
    return operationCreator(params, parentQuery, options, name);
};
const throwUnsupportedOperation = (name) => {
    throw new Error(`Unsupported operation: ${name}`);
};
const containsOperation = (query, options) => {
    for (const key in query) {
        if (options.operations.hasOwnProperty(key) || key.charAt(0) === "$")
            return true;
***REMOVED***
    return false;
};
const createNestedOperation = (keyPath, nestedQuery, parentKey, owneryQuery, options) => {
    if (containsOperation(nestedQuery, options)) {
        const [selfOperations, nestedOperations] = createQueryOperations(nestedQuery, parentKey, options);
        if (nestedOperations.length) {
            throw new Error(`Property queries must contain only operations, or exact objects.`);
    ***REMOVED***
        return new NestedOperation(keyPath, nestedQuery, owneryQuery, options, selfOperations);
***REMOVED***
    return new NestedOperation(keyPath, nestedQuery, owneryQuery, options, [
        new EqualsOperation(nestedQuery, owneryQuery, options)
    ]);
};
const createQueryOperation = (query, owneryQuery = null, { compare, operations } = {}) => {
    const options = {
        compare: compare || equals,
        operations: Object.assign({}, operations || {})
***REMOVED***;
    const [selfOperations, nestedOperations] = createQueryOperations(query, null, options);
    const ops = [];
    if (selfOperations.length) {
        ops.push(new NestedOperation([], query, owneryQuery, options, selfOperations));
***REMOVED***
    ops.push(...nestedOperations);
    if (ops.length === 1) {
        return ops[0];
***REMOVED***
    return new QueryOperation(query, owneryQuery, options, ops);
};
const createQueryOperations = (query, parentKey, options) => {
    const selfOperations = [];
    const nestedOperations = [];
    if (!isVanillaObject(query)) {
        selfOperations.push(new EqualsOperation(query, query, options));
        return [selfOperations, nestedOperations];
***REMOVED***
    for (const key in query) {
        if (options.operations.hasOwnProperty(key)) {
            const op = createNamedOperation(key, query[key], query, options);
            if (op) {
                if (!op.propop && parentKey && !options.operations[parentKey]) {
                    throw new Error(`Malformed query. ${key} cannot be matched against property.`);
            ***REMOVED***
        ***REMOVED***
            // probably just a flag for another operation (like $options)
            if (op != null) {
                selfOperations.push(op);
        ***REMOVED***
    ***REMOVED***
        else if (key.charAt(0) === "$") {
            throwUnsupportedOperation(key);
    ***REMOVED***
        else {
            nestedOperations.push(createNestedOperation(key.split("."), query[key], key, query, options));
    ***REMOVED***
***REMOVED***
    return [selfOperations, nestedOperations];
};
const createOperationTester = (operation) => (item, key, owner) => {
    operation.reset();
    operation.next(item, key, owner);
    return operation.keep;
};
const createQueryTester = (query, options = {}) => {
    return createOperationTester(createQueryOperation(query, null, options));
};

class $Ne extends BaseOperation {
    constructor() {
        super(...arguments);
        this.propop = true;
***REMOVED***
    init() {
        this._test = createTester(this.params, this.options.compare);
***REMOVED***
    reset() {
        super.reset();
        this.keep = true;
***REMOVED***
    next(item) {
        if (this._test(item)) {
            this.done = true;
            this.keep = false;
    ***REMOVED***
***REMOVED***
}
// https://docs.mongodb.com/manual/reference/operator/query/elemMatch/
class $ElemMatch extends BaseOperation {
    constructor() {
        super(...arguments);
        this.propop = true;
***REMOVED***
    init() {
        if (!this.params || typeof this.params !== "object") {
            throw new Error(`Malformed query. $elemMatch must by an object.`);
    ***REMOVED***
        this._queryOperation = createQueryOperation(this.params, this.owneryQuery, this.options);
***REMOVED***
    reset() {
        super.reset();
        this._queryOperation.reset();
***REMOVED***
    next(item) {
        if (isArray(item)) {
            for (let i = 0, { length } = item; i < length; i++) {
                // reset query operation since item being tested needs to pass _all_ query
                // operations for it to be a success
                this._queryOperation.reset();
                const child = item[i];
                this._queryOperation.next(child, i, item, false);
                this.keep = this.keep || this._queryOperation.keep;
        ***REMOVED***
            this.done = true;
    ***REMOVED***
        else {
            this.done = false;
            this.keep = false;
    ***REMOVED***
***REMOVED***
}
class $Not extends BaseOperation {
    constructor() {
        super(...arguments);
        this.propop = true;
***REMOVED***
    init() {
        this._queryOperation = createQueryOperation(this.params, this.owneryQuery, this.options);
***REMOVED***
    reset() {
        super.reset();
        this._queryOperation.reset();
***REMOVED***
    next(item, key, owner, root) {
        this._queryOperation.next(item, key, owner, root);
        this.done = this._queryOperation.done;
        this.keep = !this._queryOperation.keep;
***REMOVED***
}
class $Size extends BaseOperation {
    constructor() {
        super(...arguments);
        this.propop = true;
***REMOVED***
    init() { }
    next(item) {
        if (isArray(item) && item.length === this.params) {
            this.done = true;
            this.keep = true;
    ***REMOVED***
        // if (parent && parent.length === this.params) {
        //   this.done = true;
        //   this.keep = true;
        // }
***REMOVED***
}
const assertGroupNotEmpty = (values) => {
    if (values.length === 0) {
        throw new Error(`$and/$or/$nor must be a nonempty array`);
***REMOVED***
};
class $Or extends BaseOperation {
    constructor() {
        super(...arguments);
        this.propop = false;
***REMOVED***
    init() {
        assertGroupNotEmpty(this.params);
        this._ops = this.params.map(op => createQueryOperation(op, null, this.options));
***REMOVED***
    reset() {
        this.done = false;
        this.keep = false;
        for (let i = 0, { length } = this._ops; i < length; i++) {
            this._ops[i].reset();
    ***REMOVED***
***REMOVED***
    next(item, key, owner) {
        let done = false;
        let success = false;
        for (let i = 0, { length } = this._ops; i < length; i++) {
            const op = this._ops[i];
            op.next(item, key, owner);
            if (op.keep) {
                done = true;
                success = op.keep;
                break;
        ***REMOVED***
    ***REMOVED***
        this.keep = success;
        this.done = done;
***REMOVED***
}
class $Nor extends $Or {
    constructor() {
        super(...arguments);
        this.propop = false;
***REMOVED***
    next(item, key, owner) {
        super.next(item, key, owner);
        this.keep = !this.keep;
***REMOVED***
}
class $In extends BaseOperation {
    constructor() {
        super(...arguments);
        this.propop = true;
***REMOVED***
    init() {
        this._testers = this.params.map(value => {
            if (containsOperation(value, this.options)) {
                throw new Error(`cannot nest $ under ${this.name.toLowerCase()}`);
        ***REMOVED***
            return createTester(value, this.options.compare);
        ***REMOVED***
***REMOVED***
    next(item, key, owner) {
        let done = false;
        let success = false;
        for (let i = 0, { length } = this._testers; i < length; i++) {
            const test = this._testers[i];
            if (test(item)) {
                done = true;
                success = true;
                break;
        ***REMOVED***
    ***REMOVED***
        this.keep = success;
        this.done = done;
***REMOVED***
}
class $Nin extends BaseOperation {
    constructor(params, ownerQuery, options, name) {
        super(params, ownerQuery, options, name);
        this.propop = true;
        this._in = new $In(params, ownerQuery, options, name);
***REMOVED***
    next(item, key, owner, root) {
        this._in.next(item, key, owner);
        if (isArray(owner) && !root) {
            if (this._in.keep) {
                this.keep = false;
                this.done = true;
        ***REMOVED***
            else if (key == owner.length - 1) {
                this.keep = true;
                this.done = true;
        ***REMOVED***
    ***REMOVED***
        else {
            this.keep = !this._in.keep;
            this.done = true;
    ***REMOVED***
***REMOVED***
    reset() {
        super.reset();
        this._in.reset();
***REMOVED***
}
class $Exists extends BaseOperation {
    constructor() {
        super(...arguments);
        this.propop = true;
***REMOVED***
    next(item, key, owner) {
        if (owner.hasOwnProperty(key) === this.params) {
            this.done = true;
            this.keep = true;
    ***REMOVED***
***REMOVED***
}
class $And extends NamedGroupOperation {
    constructor(params, owneryQuery, options, name) {
        super(params, owneryQuery, options, params.map(query => createQueryOperation(query, owneryQuery, options)), name);
        this.propop = false;
        assertGroupNotEmpty(params);
***REMOVED***
    next(item, key, owner, root) {
        this.childrenNext(item, key, owner, root);
***REMOVED***
}
class $All extends NamedGroupOperation {
    constructor(params, owneryQuery, options, name) {
        super(params, owneryQuery, options, params.map(query => createQueryOperation(query, owneryQuery, options)), name);
        this.propop = true;
***REMOVED***
    next(item, key, owner, root) {
        this.childrenNext(item, key, owner, root);
***REMOVED***
}
const $eq = (params, owneryQuery, options) => new EqualsOperation(params, owneryQuery, options);
const $ne = (params, owneryQuery, options, name) => new $Ne(params, owneryQuery, options, name);
const $or = (params, owneryQuery, options, name) => new $Or(params, owneryQuery, options, name);
const $nor = (params, owneryQuery, options, name) => new $Nor(params, owneryQuery, options, name);
const $elemMatch = (params, owneryQuery, options, name) => new $ElemMatch(params, owneryQuery, options, name);
const $nin = (params, owneryQuery, options, name) => new $Nin(params, owneryQuery, options, name);
const $in = (params, owneryQuery, options, name) => {
    return new $In(params, owneryQuery, options, name);
};
const $lt = numericalOperation(params => b => b < params);
const $lte = numericalOperation(params => b => b <= params);
const $gt = numericalOperation(params => b => b > params);
const $gte = numericalOperation(params => b => b >= params);
const $mod = ([mod, equalsValue], owneryQuery, options) => new EqualsOperation(b => comparable(b) % mod === equalsValue, owneryQuery, options);
const $exists = (params, owneryQuery, options, name) => new $Exists(params, owneryQuery, options, name);
const $regex = (pattern, owneryQuery, options) => new EqualsOperation(new RegExp(pattern, owneryQuery.$options), owneryQuery, options);
const $not = (params, owneryQuery, options, name) => new $Not(params, owneryQuery, options, name);
const typeAliases = {
    number: v => typeof v === "number",
    string: v => typeof v === "string",
    bool: v => typeof v === "boolean",
    array: v => Array.isArray(v),
    null: v => v === null,
    timestamp: v => v instanceof Date
};
const $type = (clazz, owneryQuery, options) => new EqualsOperation(b => {
    if (typeof clazz === "string") {
        if (!typeAliases[clazz]) {
            throw new Error(`Type alias does not exist`);
    ***REMOVED***
        return typeAliases[clazz](b);
***REMOVED***
    return b != null ? b instanceof clazz || b.constructor === clazz : false;
}, owneryQuery, options);
const $and = (params, ownerQuery, options, name) => new $And(params, ownerQuery, options, name);
const $all = (params, ownerQuery, options, name) => new $All(params, ownerQuery, options, name);
const $size = (params, ownerQuery, options) => new $Size(params, ownerQuery, options, "$size");
const $options = () => null;
const $where = (params, ownerQuery, options) => {
    let test;
    if (isFunction(params)) {
        test = params;
***REMOVED***
    else if (!process.env.CSP_ENABLED) {
        test = new Function("obj", "return " + params);
***REMOVED***
    else {
        throw new Error(`In CSP mode, sift does not support strings in "$where" condition`);
***REMOVED***
    return new EqualsOperation(b => test.bind(b)(b), ownerQuery, options);
};

var defaultOperations = /*#__PURE__*/Object.freeze({
    __proto__: null,
    $Size: $Size,
    $eq: $eq,
    $ne: $ne,
    $or: $or,
    $nor: $nor,
    $elemMatch: $elemMatch,
    $nin: $nin,
    $in: $in,
    $lt: $lt,
    $lte: $lte,
    $gt: $gt,
    $gte: $gte,
    $mod: $mod,
    $exists: $exists,
    $regex: $regex,
    $not: $not,
    $type: $type,
    $and: $and,
    $all: $all,
    $size: $size,
    $options: $options,
    $where: $where
***REMOVED***

const createDefaultQueryOperation = (query, ownerQuery, { compare, operations } = {}) => {
    return createQueryOperation(query, ownerQuery, {
        compare,
        operations: Object.assign({}, defaultOperations, operations || {})
    ***REMOVED***
};
const createDefaultQueryTester = (query, options = {}) => {
    const op = createDefaultQueryOperation(query, null, options);
    return createOperationTester(op);
};

export default createDefaultQueryTester;
export { $Size, $all, $and, $elemMatch, $eq, $exists, $gt, $gte, $in, $lt, $lte, $mod, $ne, $nin, $nor, $not, $options, $or, $regex, $size, $type, $where, EqualsOperation, createDefaultQueryOperation, createEqualsOperation, createOperationTester, createQueryOperation, createQueryTester };
