// @aws-sdk/hash-node/dist-es/index.js
import { fromArrayBuffer, fromString } from "@aws-sdk/util-buffer-from";
import { toUint8Array } from "@aws-sdk/util-utf8";
import { Buffer } from "buffer";
import { createHash, createHmac } from "crypto";
export class Hash {
    constructor(algorithmIdentifier, secret) {
        this.algorithmIdentifier = algorithmIdentifier;
        this.secret = secret;
        this.reset();
***REMOVED***
    update(toHash, encoding) {
        this.hash.update(toUint8Array(castSourceData(toHash, encoding)));
***REMOVED***
    digest() {
        return Promise.resolve(this.hash.digest());
***REMOVED***
    reset() {
        this.hash = this.secret
            ? createHmac(this.algorithmIdentifier, castSourceData(this.secret))
            : createHash(this.algorithmIdentifier);
***REMOVED***
}
function castSourceData(toCast, encoding) {
    if (Buffer.isBuffer(toCast)) {
        return toCast;
***REMOVED***
    if (typeof toCast === "string") {
        return fromString(toCast, encoding);
***REMOVED***
    if (ArrayBuffer.isView(toCast)) {
        return fromArrayBuffer(toCast.buffer, toCast.byteOffset, toCast.byteLength);
***REMOVED***
    return fromArrayBuffer(toCast);
}
