// cors/lib/index.js
(function () {

  'use strict';

  var assign = require('object-assign');
  var vary = require('vary');

  var defaults = {
    origin: '*',
    methods: 'GET,HEAD,PUT,PATCH,POST,DELETE',
    preflightContinue: false,
    optionsSuccessStatus: 204
***REMOVED***;

  function isString(s) {
    return typeof s === 'string' || s instanceof String;
***REMOVED***

  function isOriginAllowed(origin, allowedOrigin) {
    if (Array.isArray(allowedOrigin)) {
      for (var i = 0; i < allowedOrigin.length; ++i) {
        if (isOriginAllowed(origin, allowedOrigin[i])) {
          return true;
    ***REMOVED***
  ***REMOVED***
      return false;
***REMOVED*** else if (isString(allowedOrigin)) {
      return origin === allowedOrigin;
***REMOVED*** else if (allowedOrigin instanceof RegExp) {
      return allowedOrigin.test(origin);
***REMOVED***
      return !!allowedOrigin;
***REMOVED***
***REMOVED***

  function configureOrigin(options, req) {
    var requestOrigin = req.headers.origin,
      headers = [],
      isAllowed;

    if (!options.origin || options.origin === '*') {
      // allow any origin
      headers.push([{
        key: 'Access-Control-Allow-Origin',
        value: '*'
  ***REMOVED***]);
***REMOVED*** else if (isString(options.origin)) {
      // fixed origin
      headers.push([{
        key: 'Access-Control-Allow-Origin',
        value: options.origin
  ***REMOVED***]);
      headers.push([{
        key: 'Vary',
        value: 'Origin'
  ***REMOVED***]);
***REMOVED***
      isAllowed = isOriginAllowed(requestOrigin, options.origin);
      // reflect origin
      headers.push([{
        key: 'Access-Control-Allow-Origin',
        value: isAllowed ? requestOrigin : false
  ***REMOVED***]);
      headers.push([{
        key: 'Vary',
        value: 'Origin'
  ***REMOVED***]);
***REMOVED***

    return headers;
***REMOVED***

  function configureMethods(options) {
    var methods = options.methods;
    if (methods.join) {
      methods = options.methods.join(','); // .methods is an array, so turn it into a string
***REMOVED***
    return {
      key: 'Access-Control-Allow-Methods',
      value: methods
***REMOVED***;
***REMOVED***

  function configureCredentials(options) {
    if (options.credentials === true) {
      return {
        key: 'Access-Control-Allow-Credentials',
        value: 'true'
  ***REMOVED***;
***REMOVED***
    return null;
***REMOVED***

  function configureAllowedHeaders(options, req) {
    var allowedHeaders = options.allowedHeaders || options.headers;
    var headers = [];

    if (!allowedHeaders) {
      allowedHeaders = req.headers['access-control-request-headers']; // .headers wasn't specified, so reflect the request headers
      headers.push([{
        key: 'Vary',
        value: 'Access-Control-Request-Headers'
  ***REMOVED***]);
***REMOVED*** else if (allowedHeaders.join) {
      allowedHeaders = allowedHeaders.join(','); // .headers is an array, so turn it into a string
***REMOVED***
    if (allowedHeaders && allowedHeaders.length) {
      headers.push([{
        key: 'Access-Control-Allow-Headers',
        value: allowedHeaders
  ***REMOVED***]);
***REMOVED***

    return headers;
***REMOVED***

  function configureExposedHeaders(options) {
    var headers = options.exposedHeaders;
    if (!headers) {
      return null;
***REMOVED*** else if (headers.join) {
      headers = headers.join(','); // .headers is an array, so turn it into a string
***REMOVED***
    if (headers && headers.length) {
      return {
        key: 'Access-Control-Expose-Headers',
        value: headers
  ***REMOVED***;
***REMOVED***
    return null;
***REMOVED***

  function configureMaxAge(options) {
    var maxAge = (typeof options.maxAge === 'number' || options.maxAge) && options.maxAge.toString()
    if (maxAge && maxAge.length) {
      return {
        key: 'Access-Control-Max-Age',
        value: maxAge
  ***REMOVED***;
***REMOVED***
    return null;
***REMOVED***

  function applyHeaders(headers, res) {
    for (var i = 0, n = headers.length; i < n; i++) {
      var header = headers[i];
      if (header) {
        if (Array.isArray(header)) {
          applyHeaders(header, res);
    ***REMOVED*** else if (header.key === 'Vary' && header.value) {
          vary(res, header.value);
    ***REMOVED*** else if (header.value) {
          res.setHeader(header.key, header.value);
    ***REMOVED***
  ***REMOVED***
***REMOVED***
***REMOVED***

  function cors(options, req, res, next) {
    var headers = [],
      method = req.method && req.method.toUpperCase && req.method.toUpperCase();

    if (method === 'OPTIONS') {
      // preflight
      headers.push(configureOrigin(options, req));
      headers.push(configureCredentials(options, req));
      headers.push(configureMethods(options, req));
      headers.push(configureAllowedHeaders(options, req));
      headers.push(configureMaxAge(options, req));
      headers.push(configureExposedHeaders(options, req));
      applyHeaders(headers, res);

      if (options.preflightContinue) {
        next();
  ***REMOVED***
        // Safari (and potentially other browsers) need content-length 0,
        //   for 204 or they just hang waiting for a body
        res.statusCode = options.optionsSuccessStatus;
        res.setHeader('Content-Length', '0');
        res.end();
  ***REMOVED***
***REMOVED***
      // actual response
      headers.push(configureOrigin(options, req));
      headers.push(configureCredentials(options, req));
      headers.push(configureExposedHeaders(options, req));
      applyHeaders(headers, res);
      next();
***REMOVED***
***REMOVED***

  function middlewareWrapper(o) {
    // if options are static (either via defaults or custom options passed in), wrap in a function
    var optionsCallback = null;
    if (typeof o === 'function') {
      optionsCallback = o;
***REMOVED***
      optionsCallback = function (req, cb) {
        cb(null, o);
  ***REMOVED***;
***REMOVED***

    return function corsMiddleware(req, res, next) {
      optionsCallback(req, function (err, options) {
        if (err) {
          next(err);
    ***REMOVED***
          var corsOptions = assign({}, defaults, options);
          var originCallback = null;
          if (corsOptions.origin && typeof corsOptions.origin === 'function') {
            originCallback = corsOptions.origin;
      ***REMOVED*** else if (corsOptions.origin) {
            originCallback = function (origin, cb) {
              cb(null, corsOptions.origin);
        ***REMOVED***;
      ***REMOVED***

          if (originCallback) {
            originCallback(req.headers.origin, function (err2, origin) {
              if (err2 || !origin) {
                next(err2);
          ***REMOVED***
                corsOptions.origin = origin;
                cors(corsOptions, req, res, next);
          ***REMOVED***
            ***REMOVED***
      ***REMOVED***
            next();
      ***REMOVED***
    ***REMOVED***
      ***REMOVED***
***REMOVED***;
***REMOVED***

  // can pass either an options hash, an options delegate, or nothing
  module.exports = middlewareWrapper;

}());
