// @aws-sdk/client-cognito-identity/dist-es/models/models_0.js
import { CognitoIdentityServiceException as __BaseException } from "./CognitoIdentityServiceException";
export var AmbiguousRoleResolutionType;
(function (AmbiguousRoleResolutionType) {
    AmbiguousRoleResolutionType["AUTHENTICATED_ROLE"] = "AuthenticatedRole";
    AmbiguousRoleResolutionType["DENY"] = "Deny";
})(AmbiguousRoleResolutionType || (AmbiguousRoleResolutionType = {}));
export class InternalErrorException extends __BaseException {
    constructor(opts) {
        super({
            name: "InternalErrorException",
            $fault: "server",
            ...opts,
        ***REMOVED***
        this.name = "InternalErrorException";
        this.$fault = "server";
        Object.setPrototypeOf(this, InternalErrorException.prototype);
***REMOVED***
}
export class InvalidParameterException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidParameterException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "InvalidParameterException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidParameterException.prototype);
***REMOVED***
}
export class LimitExceededException extends __BaseException {
    constructor(opts) {
        super({
            name: "LimitExceededException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "LimitExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, LimitExceededException.prototype);
***REMOVED***
}
export class NotAuthorizedException extends __BaseException {
    constructor(opts) {
        super({
            name: "NotAuthorizedException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "NotAuthorizedException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NotAuthorizedException.prototype);
***REMOVED***
}
export class ResourceConflictException extends __BaseException {
    constructor(opts) {
        super({
            name: "ResourceConflictException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "ResourceConflictException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ResourceConflictException.prototype);
***REMOVED***
}
export class TooManyRequestsException extends __BaseException {
    constructor(opts) {
        super({
            name: "TooManyRequestsException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "TooManyRequestsException";
        this.$fault = "client";
        Object.setPrototypeOf(this, TooManyRequestsException.prototype);
***REMOVED***
}
export var ErrorCode;
(function (ErrorCode) {
    ErrorCode["ACCESS_DENIED"] = "AccessDenied";
    ErrorCode["INTERNAL_SERVER_ERROR"] = "InternalServerError";
})(ErrorCode || (ErrorCode = {}));
export class ResourceNotFoundException extends __BaseException {
    constructor(opts) {
        super({
            name: "ResourceNotFoundException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "ResourceNotFoundException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ResourceNotFoundException.prototype);
***REMOVED***
}
export class ExternalServiceException extends __BaseException {
    constructor(opts) {
        super({
            name: "ExternalServiceException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "ExternalServiceException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ExternalServiceException.prototype);
***REMOVED***
}
export class InvalidIdentityPoolConfigurationException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidIdentityPoolConfigurationException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "InvalidIdentityPoolConfigurationException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidIdentityPoolConfigurationException.prototype);
***REMOVED***
}
export var MappingRuleMatchType;
(function (MappingRuleMatchType) {
    MappingRuleMatchType["CONTAINS"] = "Contains";
    MappingRuleMatchType["EQUALS"] = "Equals";
    MappingRuleMatchType["NOT_EQUAL"] = "NotEqual";
    MappingRuleMatchType["STARTS_WITH"] = "StartsWith";
})(MappingRuleMatchType || (MappingRuleMatchType = {}));
export var RoleMappingType;
(function (RoleMappingType) {
    RoleMappingType["RULES"] = "Rules";
    RoleMappingType["TOKEN"] = "Token";
})(RoleMappingType || (RoleMappingType = {}));
export class DeveloperUserAlreadyRegisteredException extends __BaseException {
    constructor(opts) {
        super({
            name: "DeveloperUserAlreadyRegisteredException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "DeveloperUserAlreadyRegisteredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, DeveloperUserAlreadyRegisteredException.prototype);
***REMOVED***
}
export class ConcurrentModificationException extends __BaseException {
    constructor(opts) {
        super({
            name: "ConcurrentModificationException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "ConcurrentModificationException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ConcurrentModificationException.prototype);
***REMOVED***
}
export const CognitoIdentityProviderFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const CreateIdentityPoolInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const IdentityPoolFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const DeleteIdentitiesInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const UnprocessedIdentityIdFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const DeleteIdentitiesResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const DeleteIdentityPoolInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const DescribeIdentityInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const IdentityDescriptionFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const DescribeIdentityPoolInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const GetCredentialsForIdentityInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const CredentialsFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const GetCredentialsForIdentityResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const GetIdInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const GetIdResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const GetIdentityPoolRolesInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const MappingRuleFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const RulesConfigurationTypeFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const RoleMappingFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const GetIdentityPoolRolesResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const GetOpenIdTokenInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const GetOpenIdTokenResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const GetOpenIdTokenForDeveloperIdentityInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const GetOpenIdTokenForDeveloperIdentityResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const GetPrincipalTagAttributeMapInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const GetPrincipalTagAttributeMapResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const ListIdentitiesInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const ListIdentitiesResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const ListIdentityPoolsInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const IdentityPoolShortDescriptionFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const ListIdentityPoolsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const ListTagsForResourceInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const ListTagsForResourceResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const LookupDeveloperIdentityInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const LookupDeveloperIdentityResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const MergeDeveloperIdentitiesInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const MergeDeveloperIdentitiesResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const SetIdentityPoolRolesInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const SetPrincipalTagAttributeMapInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const SetPrincipalTagAttributeMapResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const TagResourceInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const TagResourceResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const UnlinkDeveloperIdentityInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const UnlinkIdentityInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const UntagResourceInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const UntagResourceResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
