// @aws-sdk/client-sts/dist-es/models/models_0.js
import { STSServiceException as __BaseException } from "./STSServiceException";
export class ExpiredTokenException extends __BaseException {
    constructor(opts) {
        super({
            name: "ExpiredTokenException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "ExpiredTokenException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ExpiredTokenException.prototype);
***REMOVED***
}
export class MalformedPolicyDocumentException extends __BaseException {
    constructor(opts) {
        super({
            name: "MalformedPolicyDocumentException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "MalformedPolicyDocumentException";
        this.$fault = "client";
        Object.setPrototypeOf(this, MalformedPolicyDocumentException.prototype);
***REMOVED***
}
export class PackedPolicyTooLargeException extends __BaseException {
    constructor(opts) {
        super({
            name: "PackedPolicyTooLargeException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "PackedPolicyTooLargeException";
        this.$fault = "client";
        Object.setPrototypeOf(this, PackedPolicyTooLargeException.prototype);
***REMOVED***
}
export class RegionDisabledException extends __BaseException {
    constructor(opts) {
        super({
            name: "RegionDisabledException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "RegionDisabledException";
        this.$fault = "client";
        Object.setPrototypeOf(this, RegionDisabledException.prototype);
***REMOVED***
}
export class IDPRejectedClaimException extends __BaseException {
    constructor(opts) {
        super({
            name: "IDPRejectedClaimException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "IDPRejectedClaimException";
        this.$fault = "client";
        Object.setPrototypeOf(this, IDPRejectedClaimException.prototype);
***REMOVED***
}
export class InvalidIdentityTokenException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidIdentityTokenException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "InvalidIdentityTokenException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidIdentityTokenException.prototype);
***REMOVED***
}
export class IDPCommunicationErrorException extends __BaseException {
    constructor(opts) {
        super({
            name: "IDPCommunicationErrorException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "IDPCommunicationErrorException";
        this.$fault = "client";
        Object.setPrototypeOf(this, IDPCommunicationErrorException.prototype);
***REMOVED***
}
export class InvalidAuthorizationMessageException extends __BaseException {
    constructor(opts) {
        super({
            name: "InvalidAuthorizationMessageException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "InvalidAuthorizationMessageException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidAuthorizationMessageException.prototype);
***REMOVED***
}
export const AssumedRoleUserFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const PolicyDescriptorTypeFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const TagFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const AssumeRoleRequestFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const CredentialsFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const AssumeRoleResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const AssumeRoleWithSAMLRequestFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const AssumeRoleWithSAMLResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const AssumeRoleWithWebIdentityRequestFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const AssumeRoleWithWebIdentityResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const DecodeAuthorizationMessageRequestFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const DecodeAuthorizationMessageResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const GetAccessKeyInfoRequestFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const GetAccessKeyInfoResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const GetCallerIdentityRequestFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const GetCallerIdentityResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const GetFederationTokenRequestFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const FederatedUserFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const GetFederationTokenResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const GetSessionTokenRequestFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
export const GetSessionTokenResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
