// semver/classes/comparator.js
const ANY = Symbol('SemVer ANY')
// hoisted class for cyclic dependency
class Comparator {
  static get ANY () {
    return ANY
***REMOVED***

  constructor (comp, options) {
    options = parseOptions(options)

    if (comp instanceof Comparator) {
      if (comp.loose === !!options.loose) {
        return comp
  ***REMOVED***
        comp = comp.value
  ***REMOVED***
***REMOVED***

    debug('comparator', comp, options)
    this.options = options
    this.loose = !!options.loose
    this.parse(comp)

    if (this.semver === ANY) {
      this.value = ''
***REMOVED***
      this.value = this.operator + this.semver.version
***REMOVED***

    debug('comp', this)
***REMOVED***

  parse (comp) {
    const r = this.options.loose ? re[t.COMPARATORLOOSE] : re[t.COMPARATOR]
    const m = comp.match(r)

    if (!m) {
      throw new TypeError(`Invalid comparator: ${comp}`)
***REMOVED***

    this.operator = m[1] !== undefined ? m[1] : ''
    if (this.operator === '=') {
      this.operator = ''
***REMOVED***

    // if it literally is just '>' or '' then allow anything.
    if (!m[2]) {
      this.semver = ANY
***REMOVED***
      this.semver = new SemVer(m[2], this.options.loose)
***REMOVED***
***REMOVED***

  toString () {
    return this.value
***REMOVED***

  test (version) {
    debug('Comparator.test', version, this.options.loose)

    if (this.semver === ANY || version === ANY) {
      return true
***REMOVED***

    if (typeof version === 'string') {
    ***REMOVED***
        version = new SemVer(version, this.options)
  ***REMOVED*** catch (er) {
        return false
  ***REMOVED***
***REMOVED***

    return cmp(version, this.operator, this.semver, this.options)
***REMOVED***

  intersects (comp, options) {
    if (!(comp instanceof Comparator)) {
      throw new TypeError('a Comparator is required')
***REMOVED***

    if (!options || typeof options !== 'object') {
      options = {
        loose: !!options,
        includePrerelease: false,
  ***REMOVED***
***REMOVED***

    if (this.operator === '') {
      if (this.value === '') {
        return true
  ***REMOVED***
      return new Range(comp.value, options).test(this.value)
***REMOVED*** else if (comp.operator === '') {
      if (comp.value === '') {
        return true
  ***REMOVED***
      return new Range(this.value, options).test(comp.semver)
***REMOVED***

    const sameDirectionIncreasing =
      (this.operator === '>=' || this.operator === '>') &&
      (comp.operator === '>=' || comp.operator === '>')
    const sameDirectionDecreasing =
      (this.operator === '<=' || this.operator === '<') &&
      (comp.operator === '<=' || comp.operator === '<')
    const sameSemVer = this.semver.version === comp.semver.version
    const differentDirectionsInclusive =
      (this.operator === '>=' || this.operator === '<=') &&
      (comp.operator === '>=' || comp.operator === '<=')
    const oppositeDirectionsLessThan =
      cmp(this.semver, '<', comp.semver, options) &&
      (this.operator === '>=' || this.operator === '>') &&
        (comp.operator === '<=' || comp.operator === '<')
    const oppositeDirectionsGreaterThan =
      cmp(this.semver, '>', comp.semver, options) &&
      (this.operator === '<=' || this.operator === '<') &&
        (comp.operator === '>=' || comp.operator === '>')

    return (
      sameDirectionIncreasing ||
      sameDirectionDecreasing ||
      (sameSemVer && differentDirectionsInclusive) ||
      oppositeDirectionsLessThan ||
      oppositeDirectionsGreaterThan
    )
***REMOVED***
}

module.exports = Comparator

const parseOptions = require('../internal/parse-options')
const { re, t } = require('../internal/re')
const cmp = require('../functions/cmp')
const debug = require('../internal/debug')
const SemVer = require('./semver')
const Range = require('./range')
