// mongodb/lib/operations/find.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***
exports.FindOperation = void 0;
const error_1 = require("../error");
const read_concern_1 = require("../read_concern");
const sort_1 = require("../sort");
const utils_1 = require("../utils");
const command_1 = require("./command");
const operation_1 = require("./operation");
/** @internal */
class FindOperation extends command_1.CommandOperation {
    constructor(collection, ns, filter = {}, options = {}) {
        super(collection, options);
        this.options = options;
        this.ns = ns;
        if (typeof filter !== 'object' || Array.isArray(filter)) {
            throw new error_1.MongoInvalidArgumentError('Query filter must be a plain object or ObjectId');
    ***REMOVED***
        // If the filter is a buffer, validate that is a valid BSON document
        if (Buffer.isBuffer(filter)) {
            const objectSize = filter[0] | (filter[1] << 8) | (filter[2] << 16) | (filter[3] << 24);
            if (objectSize !== filter.length) {
                throw new error_1.MongoInvalidArgumentError(`Query filter raw message size does not match message header size [${filter.length}] != [${objectSize}]`);
        ***REMOVED***
    ***REMOVED***
        // special case passing in an ObjectId as a filter
        this.filter = filter != null && filter._bsontype === 'ObjectID' ? { _id: filter } : filter;
***REMOVED***
    execute(server, session, callback) {
        this.server = server;
        const options = this.options;
        let findCommand = makeFindCommand(this.ns, this.filter, options);
        if (this.explain) {
            findCommand = (0, utils_1.decorateWithExplain)(findCommand, this.explain);
    ***REMOVED***
        server.command(this.ns, findCommand, {
            ...this.options,
            ...this.bsonOptions,
            documentsReturnedIn: 'firstBatch',
            session
    ***REMOVED***, callback);
***REMOVED***
}
exports.FindOperation = FindOperation;
function makeFindCommand(ns, filter, options) {
    const findCommand = {
        find: ns.collection,
        filter
***REMOVED***;
    if (options.sort) {
        findCommand.sort = (0, sort_1.formatSort)(options.sort);
***REMOVED***
    if (options.projection) {
        let projection = options.projection;
        if (projection && Array.isArray(projection)) {
            projection = projection.length
                ? projection.reduce((result, field) => {
                    result[field] = 1;
                    return result;
            ***REMOVED***, {})
                : { _id: 1 };
    ***REMOVED***
        findCommand.projection = projection;
***REMOVED***
    if (options.hint) {
        findCommand.hint = (0, utils_1.normalizeHintField)(options.hint);
***REMOVED***
    if (typeof options.skip === 'number') {
        findCommand.skip = options.skip;
***REMOVED***
    if (typeof options.limit === 'number') {
        if (options.limit < 0) {
            findCommand.limit = -options.limit;
            findCommand.singleBatch = true;
    ***REMOVED***
        else {
            findCommand.limit = options.limit;
    ***REMOVED***
***REMOVED***
    if (typeof options.batchSize === 'number') {
        if (options.batchSize < 0) {
            if (options.limit &&
                options.limit !== 0 &&
                Math.abs(options.batchSize) < Math.abs(options.limit)) {
                findCommand.limit = -options.batchSize;
        ***REMOVED***
            findCommand.singleBatch = true;
    ***REMOVED***
        else {
            findCommand.batchSize = options.batchSize;
    ***REMOVED***
***REMOVED***
    if (typeof options.singleBatch === 'boolean') {
        findCommand.singleBatch = options.singleBatch;
***REMOVED***
    // we check for undefined specifically here to allow falsy values
    // eslint-disable-next-line no-restricted-syntax
    if (options.comment !== undefined) {
        findCommand.comment = options.comment;
***REMOVED***
    if (typeof options.maxTimeMS === 'number') {
        findCommand.maxTimeMS = options.maxTimeMS;
***REMOVED***
    const readConcern = read_concern_1.ReadConcern.fromOptions(options);
    if (readConcern) {
        findCommand.readConcern = readConcern.toJSON();
***REMOVED***
    if (options.max) {
        findCommand.max = options.max;
***REMOVED***
    if (options.min) {
        findCommand.min = options.min;
***REMOVED***
    if (typeof options.returnKey === 'boolean') {
        findCommand.returnKey = options.returnKey;
***REMOVED***
    if (typeof options.showRecordId === 'boolean') {
        findCommand.showRecordId = options.showRecordId;
***REMOVED***
    if (typeof options.tailable === 'boolean') {
        findCommand.tailable = options.tailable;
***REMOVED***
    if (typeof options.oplogReplay === 'boolean') {
        findCommand.oplogReplay = options.oplogReplay;
***REMOVED***
    if (typeof options.timeout === 'boolean') {
        findCommand.noCursorTimeout = !options.timeout;
***REMOVED***
    else if (typeof options.noCursorTimeout === 'boolean') {
        findCommand.noCursorTimeout = options.noCursorTimeout;
***REMOVED***
    if (typeof options.awaitData === 'boolean') {
        findCommand.awaitData = options.awaitData;
***REMOVED***
    if (typeof options.allowPartialResults === 'boolean') {
        findCommand.allowPartialResults = options.allowPartialResults;
***REMOVED***
    if (options.collation) {
        findCommand.collation = options.collation;
***REMOVED***
    if (typeof options.allowDiskUse === 'boolean') {
        findCommand.allowDiskUse = options.allowDiskUse;
***REMOVED***
    if (options.let) {
        findCommand.let = options.let;
***REMOVED***
    return findCommand;
}
(0, operation_1.defineAspects)(FindOperation, [
    operation_1.Aspect.READ_OPERATION,
    operation_1.Aspect.RETRYABLE,
    operation_1.Aspect.EXPLAINABLE,
    operation_1.Aspect.CURSOR_CREATING
]);
