// @aws-sdk/client-cognito-identity/dist-cjs/models/models_0.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true ***REMOVED***
exports.LookupDeveloperIdentityResponseFilterSensitiveLog = exports.LookupDeveloperIdentityInputFilterSensitiveLog = exports.ListTagsForResourceResponseFilterSensitiveLog = exports.ListTagsForResourceInputFilterSensitiveLog = exports.ListIdentityPoolsResponseFilterSensitiveLog = exports.IdentityPoolShortDescriptionFilterSensitiveLog = exports.ListIdentityPoolsInputFilterSensitiveLog = exports.ListIdentitiesResponseFilterSensitiveLog = exports.ListIdentitiesInputFilterSensitiveLog = exports.GetPrincipalTagAttributeMapResponseFilterSensitiveLog = exports.GetPrincipalTagAttributeMapInputFilterSensitiveLog = exports.GetOpenIdTokenForDeveloperIdentityResponseFilterSensitiveLog = exports.GetOpenIdTokenForDeveloperIdentityInputFilterSensitiveLog = exports.GetOpenIdTokenResponseFilterSensitiveLog = exports.GetOpenIdTokenInputFilterSensitiveLog = exports.GetIdentityPoolRolesResponseFilterSensitiveLog = exports.RoleMappingFilterSensitiveLog = exports.RulesConfigurationTypeFilterSensitiveLog = exports.MappingRuleFilterSensitiveLog = exports.GetIdentityPoolRolesInputFilterSensitiveLog = exports.GetIdResponseFilterSensitiveLog = exports.GetIdInputFilterSensitiveLog = exports.GetCredentialsForIdentityResponseFilterSensitiveLog = exports.CredentialsFilterSensitiveLog = exports.GetCredentialsForIdentityInputFilterSensitiveLog = exports.DescribeIdentityPoolInputFilterSensitiveLog = exports.IdentityDescriptionFilterSensitiveLog = exports.DescribeIdentityInputFilterSensitiveLog = exports.DeleteIdentityPoolInputFilterSensitiveLog = exports.DeleteIdentitiesResponseFilterSensitiveLog = exports.UnprocessedIdentityIdFilterSensitiveLog = exports.DeleteIdentitiesInputFilterSensitiveLog = exports.IdentityPoolFilterSensitiveLog = exports.CreateIdentityPoolInputFilterSensitiveLog = exports.CognitoIdentityProviderFilterSensitiveLog = exports.ConcurrentModificationException = exports.DeveloperUserAlreadyRegisteredException = exports.RoleMappingType = exports.MappingRuleMatchType = exports.InvalidIdentityPoolConfigurationException = exports.ExternalServiceException = exports.ResourceNotFoundException = exports.ErrorCode = exports.TooManyRequestsException = exports.ResourceConflictException = exports.NotAuthorizedException = exports.LimitExceededException = exports.InvalidParameterException = exports.InternalErrorException = exports.AmbiguousRoleResolutionType = void 0;
exports.UntagResourceResponseFilterSensitiveLog = exports.UntagResourceInputFilterSensitiveLog = exports.UnlinkIdentityInputFilterSensitiveLog = exports.UnlinkDeveloperIdentityInputFilterSensitiveLog = exports.TagResourceResponseFilterSensitiveLog = exports.TagResourceInputFilterSensitiveLog = exports.SetPrincipalTagAttributeMapResponseFilterSensitiveLog = exports.SetPrincipalTagAttributeMapInputFilterSensitiveLog = exports.SetIdentityPoolRolesInputFilterSensitiveLog = exports.MergeDeveloperIdentitiesResponseFilterSensitiveLog = exports.MergeDeveloperIdentitiesInputFilterSensitiveLog = void 0;
const CognitoIdentityServiceException_1 = require("./CognitoIdentityServiceException");
var AmbiguousRoleResolutionType;
(function (AmbiguousRoleResolutionType) {
    AmbiguousRoleResolutionType["AUTHENTICATED_ROLE"] = "AuthenticatedRole";
    AmbiguousRoleResolutionType["DENY"] = "Deny";
})(AmbiguousRoleResolutionType = exports.AmbiguousRoleResolutionType || (exports.AmbiguousRoleResolutionType = {}));
class InternalErrorException extends CognitoIdentityServiceException_1.CognitoIdentityServiceException {
    constructor(opts) {
        super({
            name: "InternalErrorException",
            $fault: "server",
            ...opts,
        ***REMOVED***
        this.name = "InternalErrorException";
        this.$fault = "server";
        Object.setPrototypeOf(this, InternalErrorException.prototype);
***REMOVED***
}
exports.InternalErrorException = InternalErrorException;
class InvalidParameterException extends CognitoIdentityServiceException_1.CognitoIdentityServiceException {
    constructor(opts) {
        super({
            name: "InvalidParameterException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "InvalidParameterException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidParameterException.prototype);
***REMOVED***
}
exports.InvalidParameterException = InvalidParameterException;
class LimitExceededException extends CognitoIdentityServiceException_1.CognitoIdentityServiceException {
    constructor(opts) {
        super({
            name: "LimitExceededException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "LimitExceededException";
        this.$fault = "client";
        Object.setPrototypeOf(this, LimitExceededException.prototype);
***REMOVED***
}
exports.LimitExceededException = LimitExceededException;
class NotAuthorizedException extends CognitoIdentityServiceException_1.CognitoIdentityServiceException {
    constructor(opts) {
        super({
            name: "NotAuthorizedException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "NotAuthorizedException";
        this.$fault = "client";
        Object.setPrototypeOf(this, NotAuthorizedException.prototype);
***REMOVED***
}
exports.NotAuthorizedException = NotAuthorizedException;
class ResourceConflictException extends CognitoIdentityServiceException_1.CognitoIdentityServiceException {
    constructor(opts) {
        super({
            name: "ResourceConflictException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "ResourceConflictException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ResourceConflictException.prototype);
***REMOVED***
}
exports.ResourceConflictException = ResourceConflictException;
class TooManyRequestsException extends CognitoIdentityServiceException_1.CognitoIdentityServiceException {
    constructor(opts) {
        super({
            name: "TooManyRequestsException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "TooManyRequestsException";
        this.$fault = "client";
        Object.setPrototypeOf(this, TooManyRequestsException.prototype);
***REMOVED***
}
exports.TooManyRequestsException = TooManyRequestsException;
var ErrorCode;
(function (ErrorCode) {
    ErrorCode["ACCESS_DENIED"] = "AccessDenied";
    ErrorCode["INTERNAL_SERVER_ERROR"] = "InternalServerError";
})(ErrorCode = exports.ErrorCode || (exports.ErrorCode = {}));
class ResourceNotFoundException extends CognitoIdentityServiceException_1.CognitoIdentityServiceException {
    constructor(opts) {
        super({
            name: "ResourceNotFoundException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "ResourceNotFoundException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ResourceNotFoundException.prototype);
***REMOVED***
}
exports.ResourceNotFoundException = ResourceNotFoundException;
class ExternalServiceException extends CognitoIdentityServiceException_1.CognitoIdentityServiceException {
    constructor(opts) {
        super({
            name: "ExternalServiceException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "ExternalServiceException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ExternalServiceException.prototype);
***REMOVED***
}
exports.ExternalServiceException = ExternalServiceException;
class InvalidIdentityPoolConfigurationException extends CognitoIdentityServiceException_1.CognitoIdentityServiceException {
    constructor(opts) {
        super({
            name: "InvalidIdentityPoolConfigurationException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "InvalidIdentityPoolConfigurationException";
        this.$fault = "client";
        Object.setPrototypeOf(this, InvalidIdentityPoolConfigurationException.prototype);
***REMOVED***
}
exports.InvalidIdentityPoolConfigurationException = InvalidIdentityPoolConfigurationException;
var MappingRuleMatchType;
(function (MappingRuleMatchType) {
    MappingRuleMatchType["CONTAINS"] = "Contains";
    MappingRuleMatchType["EQUALS"] = "Equals";
    MappingRuleMatchType["NOT_EQUAL"] = "NotEqual";
    MappingRuleMatchType["STARTS_WITH"] = "StartsWith";
})(MappingRuleMatchType = exports.MappingRuleMatchType || (exports.MappingRuleMatchType = {}));
var RoleMappingType;
(function (RoleMappingType) {
    RoleMappingType["RULES"] = "Rules";
    RoleMappingType["TOKEN"] = "Token";
})(RoleMappingType = exports.RoleMappingType || (exports.RoleMappingType = {}));
class DeveloperUserAlreadyRegisteredException extends CognitoIdentityServiceException_1.CognitoIdentityServiceException {
    constructor(opts) {
        super({
            name: "DeveloperUserAlreadyRegisteredException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "DeveloperUserAlreadyRegisteredException";
        this.$fault = "client";
        Object.setPrototypeOf(this, DeveloperUserAlreadyRegisteredException.prototype);
***REMOVED***
}
exports.DeveloperUserAlreadyRegisteredException = DeveloperUserAlreadyRegisteredException;
class ConcurrentModificationException extends CognitoIdentityServiceException_1.CognitoIdentityServiceException {
    constructor(opts) {
        super({
            name: "ConcurrentModificationException",
            $fault: "client",
            ...opts,
        ***REMOVED***
        this.name = "ConcurrentModificationException";
        this.$fault = "client";
        Object.setPrototypeOf(this, ConcurrentModificationException.prototype);
***REMOVED***
}
exports.ConcurrentModificationException = ConcurrentModificationException;
const CognitoIdentityProviderFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.CognitoIdentityProviderFilterSensitiveLog = CognitoIdentityProviderFilterSensitiveLog;
const CreateIdentityPoolInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.CreateIdentityPoolInputFilterSensitiveLog = CreateIdentityPoolInputFilterSensitiveLog;
const IdentityPoolFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.IdentityPoolFilterSensitiveLog = IdentityPoolFilterSensitiveLog;
const DeleteIdentitiesInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.DeleteIdentitiesInputFilterSensitiveLog = DeleteIdentitiesInputFilterSensitiveLog;
const UnprocessedIdentityIdFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.UnprocessedIdentityIdFilterSensitiveLog = UnprocessedIdentityIdFilterSensitiveLog;
const DeleteIdentitiesResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.DeleteIdentitiesResponseFilterSensitiveLog = DeleteIdentitiesResponseFilterSensitiveLog;
const DeleteIdentityPoolInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.DeleteIdentityPoolInputFilterSensitiveLog = DeleteIdentityPoolInputFilterSensitiveLog;
const DescribeIdentityInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.DescribeIdentityInputFilterSensitiveLog = DescribeIdentityInputFilterSensitiveLog;
const IdentityDescriptionFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.IdentityDescriptionFilterSensitiveLog = IdentityDescriptionFilterSensitiveLog;
const DescribeIdentityPoolInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.DescribeIdentityPoolInputFilterSensitiveLog = DescribeIdentityPoolInputFilterSensitiveLog;
const GetCredentialsForIdentityInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.GetCredentialsForIdentityInputFilterSensitiveLog = GetCredentialsForIdentityInputFilterSensitiveLog;
const CredentialsFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.CredentialsFilterSensitiveLog = CredentialsFilterSensitiveLog;
const GetCredentialsForIdentityResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.GetCredentialsForIdentityResponseFilterSensitiveLog = GetCredentialsForIdentityResponseFilterSensitiveLog;
const GetIdInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.GetIdInputFilterSensitiveLog = GetIdInputFilterSensitiveLog;
const GetIdResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.GetIdResponseFilterSensitiveLog = GetIdResponseFilterSensitiveLog;
const GetIdentityPoolRolesInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.GetIdentityPoolRolesInputFilterSensitiveLog = GetIdentityPoolRolesInputFilterSensitiveLog;
const MappingRuleFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.MappingRuleFilterSensitiveLog = MappingRuleFilterSensitiveLog;
const RulesConfigurationTypeFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.RulesConfigurationTypeFilterSensitiveLog = RulesConfigurationTypeFilterSensitiveLog;
const RoleMappingFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.RoleMappingFilterSensitiveLog = RoleMappingFilterSensitiveLog;
const GetIdentityPoolRolesResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.GetIdentityPoolRolesResponseFilterSensitiveLog = GetIdentityPoolRolesResponseFilterSensitiveLog;
const GetOpenIdTokenInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.GetOpenIdTokenInputFilterSensitiveLog = GetOpenIdTokenInputFilterSensitiveLog;
const GetOpenIdTokenResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.GetOpenIdTokenResponseFilterSensitiveLog = GetOpenIdTokenResponseFilterSensitiveLog;
const GetOpenIdTokenForDeveloperIdentityInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.GetOpenIdTokenForDeveloperIdentityInputFilterSensitiveLog = GetOpenIdTokenForDeveloperIdentityInputFilterSensitiveLog;
const GetOpenIdTokenForDeveloperIdentityResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.GetOpenIdTokenForDeveloperIdentityResponseFilterSensitiveLog = GetOpenIdTokenForDeveloperIdentityResponseFilterSensitiveLog;
const GetPrincipalTagAttributeMapInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.GetPrincipalTagAttributeMapInputFilterSensitiveLog = GetPrincipalTagAttributeMapInputFilterSensitiveLog;
const GetPrincipalTagAttributeMapResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.GetPrincipalTagAttributeMapResponseFilterSensitiveLog = GetPrincipalTagAttributeMapResponseFilterSensitiveLog;
const ListIdentitiesInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.ListIdentitiesInputFilterSensitiveLog = ListIdentitiesInputFilterSensitiveLog;
const ListIdentitiesResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.ListIdentitiesResponseFilterSensitiveLog = ListIdentitiesResponseFilterSensitiveLog;
const ListIdentityPoolsInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.ListIdentityPoolsInputFilterSensitiveLog = ListIdentityPoolsInputFilterSensitiveLog;
const IdentityPoolShortDescriptionFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.IdentityPoolShortDescriptionFilterSensitiveLog = IdentityPoolShortDescriptionFilterSensitiveLog;
const ListIdentityPoolsResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.ListIdentityPoolsResponseFilterSensitiveLog = ListIdentityPoolsResponseFilterSensitiveLog;
const ListTagsForResourceInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.ListTagsForResourceInputFilterSensitiveLog = ListTagsForResourceInputFilterSensitiveLog;
const ListTagsForResourceResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.ListTagsForResourceResponseFilterSensitiveLog = ListTagsForResourceResponseFilterSensitiveLog;
const LookupDeveloperIdentityInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.LookupDeveloperIdentityInputFilterSensitiveLog = LookupDeveloperIdentityInputFilterSensitiveLog;
const LookupDeveloperIdentityResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.LookupDeveloperIdentityResponseFilterSensitiveLog = LookupDeveloperIdentityResponseFilterSensitiveLog;
const MergeDeveloperIdentitiesInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.MergeDeveloperIdentitiesInputFilterSensitiveLog = MergeDeveloperIdentitiesInputFilterSensitiveLog;
const MergeDeveloperIdentitiesResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.MergeDeveloperIdentitiesResponseFilterSensitiveLog = MergeDeveloperIdentitiesResponseFilterSensitiveLog;
const SetIdentityPoolRolesInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.SetIdentityPoolRolesInputFilterSensitiveLog = SetIdentityPoolRolesInputFilterSensitiveLog;
const SetPrincipalTagAttributeMapInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.SetPrincipalTagAttributeMapInputFilterSensitiveLog = SetPrincipalTagAttributeMapInputFilterSensitiveLog;
const SetPrincipalTagAttributeMapResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.SetPrincipalTagAttributeMapResponseFilterSensitiveLog = SetPrincipalTagAttributeMapResponseFilterSensitiveLog;
const TagResourceInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.TagResourceInputFilterSensitiveLog = TagResourceInputFilterSensitiveLog;
const TagResourceResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.TagResourceResponseFilterSensitiveLog = TagResourceResponseFilterSensitiveLog;
const UnlinkDeveloperIdentityInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.UnlinkDeveloperIdentityInputFilterSensitiveLog = UnlinkDeveloperIdentityInputFilterSensitiveLog;
const UnlinkIdentityInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.UnlinkIdentityInputFilterSensitiveLog = UnlinkIdentityInputFilterSensitiveLog;
const UntagResourceInputFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.UntagResourceInputFilterSensitiveLog = UntagResourceInputFilterSensitiveLog;
const UntagResourceResponseFilterSensitiveLog = (obj) => ({
    ...obj,
***REMOVED***
exports.UntagResourceResponseFilterSensitiveLog = UntagResourceResponseFilterSensitiveLog;
